File: web/src/store/conversations.ts
import { create } from "zustand"
import { api } from "@/lib/api"

export interface Conversation {
    id: string
    title: string
    persona_id?: string
    created_at: string
    updated_at: string
}

export interface Message {
    id: string
    conversation_id: string
    role: "user" | "assistant" | "system" | "tool"
    content: string
    thought?: string
    steps?: Array<{
        thought?: string
        action?: string
        action_input?: Record<string, unknown>
        observation?: string
        is_final_answer?: boolean
        final_answer?: string
    }>
    tool_call?: { name: string; args: Record<string, unknown> }
    created_at: string
}

interface ConversationState {
    conversations: Conversation[]
    activeConversationId: string | null
    messages: Message[]
    isLoadingConversations: boolean
    isLoadingMessages: boolean

    // Actions
    fetchConversations: () => Promise<void>
    createConversation: (title?: string) => Promise<Conversation | null>
    selectConversation: (id: string) => Promise<void>
    deleteConversation: (id: string) => Promise<void>
    renameConversation: (id: string, title: string) => Promise<void>
    addLocalMessage: (msg: Message) => void
    clearActive: () => void
}

export const useConversationStore = create<ConversationState>((set, get) => ({
    conversations: [],
    activeConversationId: null,
    messages: [],
    isLoadingConversations: false,
    isLoadingMessages: false,

    fetchConversations: async () => {
        set({ isLoadingConversations: true })
        try {
            const { data } = await api.GET("/v1/conversations")
            if (data) {
                set({
                    conversations: (data as unknown as Conversation[]) ?? [],
                })
            }
        } finally {
            set({ isLoadingConversations: false })
        }
    },

    createConversation: async (title = "New Chat") => {
        const { data } = await api.POST("/v1/conversations", {
            body: { title },
        })
        if (data) {
            const conv = data as unknown as Conversation
            set((s) => ({
                conversations: [conv, ...s.conversations],
                activeConversationId: conv.id,
                messages: [],
            }))
            return conv
        }
        return null
    },

    selectConversation: async (id: string) => {
        set({ activeConversationId: id, isLoadingMessages: true, messages: [] })
        try {
            const { data } = await api.GET("/v1/conversations/{id}/messages", {
                params: { path: { id }, query: { limit: 100 } },
            })
            if (data) {
                set({ messages: (data as unknown as Message[]) ?? [] })
            }
        } finally {
            set({ isLoadingMessages: false })
        }
    },

    deleteConversation: async (id: string) => {
        await api.DELETE("/v1/conversations/{id}", {
            params: { path: { id } },
        })
        const state = get()
        set({
            conversations: state.conversations.filter((c) => c.id !== id),
            ...(state.activeConversationId === id
                ? { activeConversationId: null, messages: [] }
                : {}),
        })
    },

    renameConversation: async (id: string, title: string) => {
        await api.PATCH("/v1/conversations/{id}", {
            params: { path: { id } },
            body: { title },
        })
        set((s) => ({
            conversations: s.conversations.map((c) =>
                c.id === id ? { ...c, title } : c
            ),
        }))
    },

    addLocalMessage: (msg: Message) => {
        set((s) => ({ messages: [...s.messages, msg] }))
    },

    clearActive: () => {
        set({ activeConversationId: null, messages: [] })
    },
}))

File: pkg/kernel/conversations.go
package kernel

import (
	"context"

	"github.com/manthysbr/auleOS/internal/core/domain"
)

// --- StrictServerInterface implementations for Conversations ---

// ListConversations implements StrictServerInterface.
func (s *Server) ListConversations(ctx context.Context, _ ListConversationsRequestObject) (ListConversationsResponseObject, error) {
	convs, err := s.convStore.ListConversations(ctx)
	if err != nil {
		s.logger.Error("failed to list conversations", "error", err)
		return ListConversations200JSONResponse{}, nil
	}

	result := make(ListConversations200JSONResponse, len(convs))
	for i, c := range convs {
		result[i] = domainConvToAPI(c)
	}

	return result, nil
}

// CreateConversation implements StrictServerInterface.
func (s *Server) CreateConversation(ctx context.Context, request CreateConversationRequestObject) (CreateConversationResponseObject, error) {
	title := "New Chat"
	if request.Body != nil && request.Body.Title != nil {
		title = *request.Body.Title
	}

	conv, err := s.convStore.CreateConversation(ctx, title)
	if err != nil {
		s.logger.Error("failed to create conversation", "error", err)
		return nil, err
	}

	return CreateConversation201JSONResponse(domainConvToAPI(conv)), nil
}

// GetConversation implements StrictServerInterface.
func (s *Server) GetConversation(ctx context.Context, request GetConversationRequestObject) (GetConversationResponseObject, error) {
	conv, err := s.convStore.GetConversation(ctx, domain.ConversationID(request.Id))
	if err != nil {
		if err == domain.ErrConversationNotFound {
			msg := "conversation not found"
			return GetConversation404JSONResponse{Error: &msg}, nil
		}
		s.logger.Error("failed to get conversation", "error", err)
		return nil, err
	}

	return GetConversation200JSONResponse(domainConvToAPI(conv)), nil
}

// DeleteConversation implements StrictServerInterface.
func (s *Server) DeleteConversation(ctx context.Context, request DeleteConversationRequestObject) (DeleteConversationResponseObject, error) {
	if err := s.convStore.DeleteConversation(ctx, domain.ConversationID(request.Id)); err != nil {
		if err == domain.ErrConversationNotFound {
			return DeleteConversation404Response{}, nil
		}
		s.logger.Error("failed to delete conversation", "error", err)
		return nil, err
	}

	return DeleteConversation204Response{}, nil
}

// UpdateConversation implements StrictServerInterface.
func (s *Server) UpdateConversation(ctx context.Context, request UpdateConversationRequestObject) (UpdateConversationResponseObject, error) {
	id := domain.ConversationID(request.Id)

	if request.Body != nil && request.Body.Title != nil {
		if err := s.convStore.UpdateTitle(ctx, id, *request.Body.Title); err != nil {
			if err == domain.ErrConversationNotFound {
				return nil, err
			}
			s.logger.Error("failed to update conversation", "error", err)
			return nil, err
		}
	}

	conv, err := s.convStore.GetConversation(ctx, id)
	if err != nil {
		return nil, err
	}

	return UpdateConversation200JSONResponse(domainConvToAPI(conv)), nil
}

// ListMessages implements StrictServerInterface.
func (s *Server) ListMessages(ctx context.Context, request ListMessagesRequestObject) (ListMessagesResponseObject, error) {
	limit := 50
	if request.Params.Limit != nil {
		limit = *request.Params.Limit
	}

	msgs, err := s.convStore.GetMessages(ctx, domain.ConversationID(request.Id), limit)
	if err != nil {
		s.logger.Error("failed to list messages", "error", err)
		return ListMessages200JSONResponse{}, nil
	}

	result := make(ListMessages200JSONResponse, len(msgs))
	for i, m := range msgs {
		result[i] = domainMsgToAPI(m)
	}

	return result, nil
}

// --- Mapping helpers ---

func domainConvToAPI(c domain.Conversation) Conversation {
	id := string(c.ID)
	title := c.Title
	createdAt := c.CreatedAt
	updatedAt := c.UpdatedAt
	conv := Conversation{
		Id:        &id,
		Title:     &title,
		CreatedAt: &createdAt,
		UpdatedAt: &updatedAt,
	}
	if c.ProjectID != nil {
		pid := string(*c.ProjectID)
		conv.ProjectId = &pid
	}
	if c.PersonaID != nil {
		pid := string(*c.PersonaID)
		conv.PersonaId = &pid
	}
	return conv
}

func domainMsgToAPI(m domain.Message) Message {
	id := string(m.ID)
	convID := string(m.ConversationID)
	role := MessageRole(m.Role)
	content := m.Content
	createdAt := m.CreatedAt

	msg := Message{
		Id:             &id,
		ConversationId: &convID,
		Role:           &role,
		Content:        &content,
		CreatedAt:      &createdAt,
	}

	if m.Thought != "" {
		msg.Thought = &m.Thought
	}

	if len(m.Steps) > 0 {
		apiSteps := make([]ReActStep, len(m.Steps))
		for j, step := range m.Steps {
			apiSteps[j] = ReActStep{}
			if step.Thought != "" {
				v := step.Thought
				apiSteps[j].Thought = &v
			}
			if step.Action != "" {
				v := step.Action
				apiSteps[j].Action = &v
			}
			if step.ActionInput != nil {
				v := step.ActionInput
				apiSteps[j].ActionInput = &v
			}
			if step.Observation != "" {
				v := step.Observation
				apiSteps[j].Observation = &v
			}
			if step.FinalAnswer != "" {
				v := step.FinalAnswer
				apiSteps[j].FinalAnswer = &v
			}
			if step.IsFinalAnswer {
				v := step.IsFinalAnswer
				apiSteps[j].IsFinalAnswer = &v
			}
		}
		msg.Steps = &apiSteps
	}

	if m.ToolCall != nil {
		msg.ToolCall = &struct {
			Args *map[string]interface{} `json:"args,omitempty"`
			Name *string                 `json:"name,omitempty"`
		}{
			Name: &m.ToolCall.Name,
			Args: &m.ToolCall.Args,
		}
	}

	return msg
}
